src/soil_funcs.cxx

```diff
@@ -63,28 +65,54 @@ extern double calc_Geff(bool use_closed_form_G, double theta1, double theta2, do
       Se = calc_Se_from_h(h_f,vg_alpha,vg_m,vg_n);
       printf("Se_f = %8.6lf,  Se_inverse = %8.6lf\n", Se_f, Se);
     }
-    
+
     // nint = number of "dh" intervals to integrate over using trapezoidal rule
-    dh = (h_f-h_i)/(double)nint;
+    dh = -1*(h_f-h_i)/(double)nint;
+    dh = dh*0.01;
+
     Geff = 0.0;
-    
+
     // integrate k(h) dh from h_i to h_f, using trapezoidal rule, with subscript
     // 1 denoting the left-hand side of the trapezoid, and 2 denoting the right-hand side
 
     Se1 = Se_i;  // could just use Se_i in next statement.  Done 4 completeness.
     K1  = calc_K_from_Se(Se1, Ksat, vg_m);
-    h2  = h_i + dh;
-    
-    for(int i=1; i<=nint; i++) {
+    h2  = h_f + dh;
+
+    while(h2<h_i) {
+
+      double prior_h2 = h2;
 
       Se2 = calc_Se_from_h(h2, vg_alpha, vg_m, vg_n);
       K2  = calc_K_from_Se(Se2, Ksat, vg_m);
-      Geff += (K1+K2)*dh/2.0;                  // trapezoidal rule
+
+      if ( (K1-K2)/K2 > 0.02 ){//if K1 disagrees with K2 by more than this fraction, then dh is made smaller
+        dh = dh*0.5;
+      }
+      else if ( (K1-K2)/K2 < 0.03 ){//but if K1 and K2 are within a certain fraction of each other, then dh is made larger
+        dh = dh*10.0;
+      }
+      else{
+        dh = dh*2.0; //but if K1 and K2 are within a certain fraction of each other, then dh is made larger
+      }
```

Comment:
can you please explain in a sentence or two why this logic is needed?

Reply:
The idea is: if K1 and K2 differ by more than some factor, then the numerical integral's trapezoid width has to be made smaller in order to ensure accuracy. If the difference between K1 and K2 is less than some factor, then the numerical integral's trapezoid width can be made wider without sacrificing much accuracy. 

The above code is however equivalent to checking if (K1-K2)/K2 <= 0.02 in the else if and removing the final else because it will never occur. So for simplicity, I updated it with that expression and removed the final else. The original idea was that we could have three different factors by which to augment dh based on the difference between K1 and K2, but it seems as if a third region isn't necessary. The factor of 2 percent is arbitrary but seems to only minimally affect infiltration values.